In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
import os
import pandas_datareader as dr
plt.style.use('seaborn-darkgrid')
import plotly.graph_objects as go
import plotly.express as px

In [2]:
os.chdir(r'/Users/marike/Documents/MARIKE/DATA SCIENCE/Final project/Marike FP/Data')

In [3]:
"""indicate in below code line (convert comment to code) the location of Speed Dating Data 
in case of checking the project from another computer"""
# os.chdir("..")

cities_path = r'WB_city_level_data_0_0.csv'
cities_df = pd.read_csv(cities_path, thousands = ',')

## 1. DATA CLEANING

### 1.1 CITIES DATASET

In [4]:
cities_df.head()

iso3c region_id country_name income_id       city_name  \
0   AFG       SAS  Afghanistan       LIC       Jalalabad   
1   AFG       SAS  Afghanistan       LIC        Kandahar   
2   AFG       SAS  Afghanistan       LIC  Mazar-E-Sharif   
3   AFG       SAS  Afghanistan       LIC           Kabul   
4   AFG       SAS  Afghanistan       LIC         Hirat    

  additional_data_annual_budget_for_waste_management_year  \
0                                                NaN        
1                                                NaN        
2                                                NaN        
3                                                NaN        
4                                                NaN        

  additional_data_annual_solid_waste_budget_year  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

   additional_data_annual_swm_budget_2017_year  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   

   additional_data_annual_swm_budget_year  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   

   additional_data_annual_waste_budget_year  ...  \
0                                       NaN  ...   
1                                       NaN  ...   
2                                       NaN  ...   
3                                       NaN  ...   
4                                       NaN  ...   

   waste_treatment_compost_percent  \
0                              NaN   
1                              NaN   
2                              NaN   
3                              NaN   
4                              NaN   

   waste_treatment_controlled_landfill_percent  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   

  waste_treatment_incineration_percent  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   

  waste_treatment_landfill_unspecified_percent  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   

  waste_treatment_open_dump_percent  waste_treatment_other_percent  \
0                             100.0                            NaN   
1                             100.0                            NaN   
2                             100.0                            NaN   
3                              75.0                            NaN   
4                             100.0                            NaN   

   waste_treatment_recycling_percent  \
0                                NaN   
1                                NaN   
2                                NaN   
3                                NaN   
4                                NaN   

   waste_treatment_sanitary_landfill_landfill_gas_system_percent  \
0                                                NaN               
1                                                NaN               
2                                                NaN               
3                                                NaN      

In [5]:
#we check whether dataset has some duplicated row:
print('Cities dataset has ', cities_df.duplicated().sum(),' duplicated rows.')

Cities dataset has  0  duplicated rows.


In [6]:
cities_df.shape

(367, 113)

In [7]:
cities_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 367 entries, 0 to 366
Columns: 113 entries, iso3c to waste_treatment_waterways_marine_percent
dtypes: float64(72), object(41)
memory usage: 324.1+ KB


According to dataset information, it is composed of 113 variables, 72 are numerical variables and 41 seem to be categorical variables. As we have to many variables for the information to be displayed, we print the details of all variables.

In [8]:
for col in cities_df.columns: 
    print(col,' , ',cities_df[col].dtypes,', null values:', cities_df[col].isna().sum())

iso3c  ,  object , null values: 0
region_id  ,  object , null values: 0
country_name  ,  object , null values: 0
income_id  ,  object , null values: 0
city_name  ,  object , null values: 0
additional_data_annual_budget_for_waste_management_year  ,  object , null values: 366
additional_data_annual_solid_waste_budget_year  ,  object , null values: 366
additional_data_annual_swm_budget_2017_year  ,  float64 , null values: 366
additional_data_annual_swm_budget_year  ,  float64 , null values: 365
additional_data_annual_waste_budget_year  ,  float64 , null values: 366
additional_data_collection_ton  ,  float64 , null values: 366
additional_data_number_of_scavengers_on_dumpsites_number  ,  float64 , null values: 366
additional_data_other_user_fees_na  ,  object , null values: 366
additional_data_swm_contract_arrangement_1_year_contract_period  ,  object , null values: 366
additional_data_swm_contract_arrangement_3_year_contract_period  ,  object , null values: 366
additional_data_total_annual

Our dataset has a length of 367 rows that correspond to cities but from the information, we see that information is missing for many variables. We check the extension of this fact.

In [9]:
#We import functions we have created to use repeated times in this project.
from MC_FP_UsefulFunctions import *

In [10]:
#we use function "show_null" to display characteristics of missing information in dataframe.

cities_null_perc = show_null(cities_df)
cities_null_perc

Null data amount  \
additional_data_annual_budget_for_waste_managem...               366   
additional_data_swm_contract_arrangement_3_year...               366   
additional_data_annual_solid_waste_budget_year                   366   
transportation_distance_km                                       366   
transportation_distance_na                                       366   
...                                                              ...   
city_name                                                          0   
income_id                                                          0   
region_id                                                          0   
iso3c                                                              0   
country_name                                                       0   

                                                    Missing data (%)     type  
additional_data_annual_budget_for_waste_managem...          99.72752   object  
additional_data_swm_contract_arrangement_3_year...          99.72752   object  
additional_data_annual_solid_waste_budget_year              99.72752   object  
transportation_distance_km                                  99.72752  float64  
transportation_distance_na                                  99.72752   object  
...                                                              ...      ...  
city_name                                                    0.00000   object  
income_id                                                    0.00000   object  
region_id                                                    0.00000   object  
iso3c                                                        0.00000   object  
country_name                                                 0.00000   object  

[113 rows x 3 columns]

Some variables have almost a 100% of missing information. When the variable has o much missing information, we disregard it because it cannot give representative information. We chech the extension of the issue in our dataset:

In [11]:
print('Number of variables with more than 90% missing information: ', len(cities_null_perc.loc[cities_null_perc['Missing data (%)']>90]), '. That is to say ',round(len(cities_null_perc.loc[cities_null_perc['Missing data (%)']>90])*100/len(cities_null_perc),2),'% of the variables')
print('Number of variables with more than 75% missing information: ', len(cities_null_perc.loc[cities_null_perc['Missing data (%)']>75]), '. That is to say ',round(len(cities_null_perc.loc[cities_null_perc['Missing data (%)']>75])*100/len(cities_null_perc),2),'% of the variables')
print('Number of variables with more than 50% missing information: ', len(cities_null_perc.loc[cities_null_perc['Missing data (%)']>50]), '. That is to say ',round(len(cities_null_perc.loc[cities_null_perc['Missing data (%)']>50])*100/len(cities_null_perc),2),'% of the variables')
print('Number of variables with more than 40% missing information: ', len(cities_null_perc.loc[cities_null_perc['Missing data (%)']>40]), '. That is to say ',round(len(cities_null_perc.loc[cities_null_perc['Missing data (%)']>40])*100/len(cities_null_perc),2),'% of the variables')

Number of variables with more than 90% missing information:  54 . That is to say  47.79 % of the variables
Number of variables with more than 75% missing information:  72 . That is to say  63.72 % of the variables
Number of variables with more than 50% missing information:  95 . That is to say  84.07 % of the variables
Number of variables with more than 40% missing information:  97 . That is to say  85.84 % of the variables


Among variables without missing information, there are at least two that we consider important to look at more detailly: the *region_id* and the *income_id* because we know that waste issue is dealt with on different way according to continents and above all economical situation so before dropping massively information, we will check if the missing information is so important for each sub geographical or economic area.

In [12]:
cities_df.region_id.unique()

array(['SAS', 'SSF', 'ECS', 'MEA', 'LCN', 'EAS', 'NAC'], dtype=object)

**Meaning of regions codes:**

***EAS*: East Asia & Pacific**

***ECS*: Europe & Central Asia**

***LCN*: Latin America & Caribbean**

***MEA*: Middle East & North Africa**

***NAC*: North America**

***SAS*: South Asia**

***SSF*: Sub-Saharan Africa**


In [13]:
cities_df.income_id.unique()

array(['LIC', 'LMC', 'UMC', 'HIC'], dtype=object)

**Meaning of income codes:**

***HIC*: High Income Country**

***UMC*: Upper & Middle Income Country**

***LMC*: Lower & Middle Income Country**

***LIC*: Low Income Country**


In [14]:
cities_df.replace({'LIC': 'LowIncome', 'LMC': 'LowMiddleIncome','UMC':'MiddleUpperIncome','HIC':'HighIncome'},inplace=True)

In [15]:
Cities_null_perc_X_region_income = pd.DataFrame()
Cities_null_perc_X_region_income['World']=cities_null_perc['Missing data (%)']

for x in cities_df.region_id.unique():
    null_perc_city_region = show_null(cities_df.loc[cities_df.region_id == x])
    Cities_null_perc_X_region_income[x]=round(null_perc_city_region['Missing data (%)'],2)
    
for x in cities_df.income_id.unique():
    null_perc_city_income = show_null(cities_df.loc[cities_df.income_id == x])
    Cities_null_perc_X_region_income[x]=round(null_perc_city_income['Missing data (%)'],2)

Cities_null_perc_X_region_income

World    SAS    SSF  \
additional_data_annual_budget_for_waste_managem...  99.72752  100.0  100.0   
additional_data_swm_contract_arrangement_3_year...  99.72752  100.0  100.0   
additional_data_annual_solid_waste_budget_year      99.72752  100.0  100.0   
transportation_distance_km                          99.72752   98.8  100.0   
transportation_distance_na                          99.72752  100.0  100.0   
...                                                      ...    ...    ...   
city_name                                            0.00000    0.0    0.0   
income_id                                            0.00000    0.0    0.0   
region_id                                            0.00000    0.0    0.0   
iso3c                                                0.00000    0.0    0.0   
country_name                                         0.00000    0.0    0.0   

                                                      ECS    MEA    LCN  \
additional_data_annual_budget_for_waste_managem...  100.0  100.0  100.0   
additional_data_swm_contract_arrangement_3_year...  100.0  100.0  100.0   
additional_data_annual_solid_waste_budget_year      100.0  100.0  100.0   
transportation_distance_km                          100.0  100.0  100.0   
transportation_distance_na                          100.0  100.0  100.0   
...                                                   ...    ...    ...   
city_name                                             0.0    0.0    0.0   
income_id                                             0.0    0.0    0.0   
region_id                                             0.0    0.0    0.0   
iso3c                                                 0.0    0.0    0.0   
country_name                                          0.0    0.0    0.0   

                                                      EAS    NAC  LowIncome  \
additional_data_annual_budget_for_waste_managem...   97.5  100.0     100.00   
additional_data_swm_contract_arrangement_3_year...   97.5  100.0     100.00   
additional_data_annual_solid_waste_budget_year       97.5  100.0     100.00   
transportation_distance_km                          100.0  100.0      98.65   
transportation_distance_na                           97.5  100.0     100.00   
...                                                   ...    ...        ...   
city_name                                             0.0    0.0       0.00   
income_id                                             0.0    0.0       0.00   
region_id                                             0.0    0.0       0.00   
iso3c                                                 0.0    0.0       0.00   
country_name                                          0.0    0.0       0.00   

                                                    LowMiddleIncome  \
additional_data_annual_budget_for_waste_managem...            99.19   
additional_data_swm_contract_arrangement_3_year...           100.00   
additional_data_annual_solid_waste_budget_year                99.19   
transportation_distance_km                                   100.00   
transportation_distance_na                                   100.00   
...                                                             ...   
city_name                                                      0.00   
income_id                                                      0.00   
region_id                                                      0.00   
iso3c                                                          0.00   
country_name                                                   0.00   

                                                    MiddleUpperIncome  \
additional_data_annual_budget_for_waste_managem...             100.00   
additional_data_swm_contract_arrangement_3_year...              98.77   
additional_data_annual_solid_waste_budget_year                 100.00   
transportation_distance_km                                     100.00   
transportation_distance_na                        

We can see that some variables have a very high percentage of missing information. **We decide to keep only variables with less than 50% of missing information except for variables with 50 to 70% missing information but for which there is a world area (defined geographically or by income) where this variable has less than 30% missing information.** 

Indeed we know that some variables will only be informed in areas where waste issue is better dealt with.

In [16]:
Notallregions_Income_HighNaNPerc=[]

for x in Cities_null_perc_X_region_income.index:
    
    if Cities_null_perc_X_region_income.at[x,'World']<70 and Cities_null_perc_X_region_income.at[x,'World']>50:
        
        for y in cities_df.region_id.unique():
            if Cities_null_perc_X_region_income.at[x,y]<30:
                Notallregions_Income_HighNaNPerc.append(x)
                
        for z in cities_df.income_id.unique():
            if Cities_null_perc_X_region_income.at[x,z]<30:
                Notallregions_Income_HighNaNPerc.append(x)
                
Notallregions_Income_HighNaNPerc

['legal_framework_master_plan_is_being_implemented_na',
 'institutional_framework_unit_enforcing_solid_waste_issues_in_the_city_such_as_illegal_dumping_or_littering_na',
 'waste_treatment_recycling_percent',
 'institutional_framework_information_system_for_solid_waste_management_na',
 'separation_breakdonw_organics_na',
 'waste_collection_cost_recovery_household_fee_amount_na',
 'separation_breakdonw_cans_and_metals_na',
 'separation_breakdonw_cans_and_metals_na',
 'separation_breakdonw_glass_na',
 'separation_breakdonw_glass_na',
 'separation_breakdonw_plastics_and_packaging_na',
 'separation_breakdonw_plastics_and_packaging_na',
 'separation_breakdonw_paper_cardboard_na',
 'separation_breakdonw_paper_cardboard_na',
 'waste_collection_coverage_total_percent_of_waste_percent_of_waste',
 'waste_collection_coverage_total_percent_of_waste_percent_of_waste']

In [17]:
Notallregions_Income_HighNaNPerc_df = pd.DataFrame({'Variable to keep':Notallregions_Income_HighNaNPerc})

In [18]:
Notallregions_Income_HighNaNPerc_df['Variable to keep'].unique()

array(['legal_framework_master_plan_is_being_implemented_na',
       'institutional_framework_unit_enforcing_solid_waste_issues_in_the_city_such_as_illegal_dumping_or_littering_na',
       'waste_treatment_recycling_percent',
       'institutional_framework_information_system_for_solid_waste_management_na',
       'separation_breakdonw_organics_na',
       'waste_collection_cost_recovery_household_fee_amount_na',
       'separation_breakdonw_cans_and_metals_na',
       'separation_breakdonw_glass_na',
       'separation_breakdonw_plastics_and_packaging_na',
       'separation_breakdonw_paper_cardboard_na',
       'waste_collection_coverage_total_percent_of_waste_percent_of_waste'],
      dtype=object)

In [19]:
ColToDrop = []
for x in Cities_null_perc_X_region_income.index:
    if Cities_null_perc_X_region_income.at[x,'World']>50:
        if x not in Notallregions_Income_HighNaNPerc_df['Variable to keep'].unique():
            ColToDrop.append(x)

ColToDrop

['additional_data_annual_budget_for_waste_management_year',
 'additional_data_swm_contract_arrangement_3_year_contract_period',
 'additional_data_annual_solid_waste_budget_year',
 'transportation_distance_km',
 'transportation_distance_na',
 'transportation_distance_total_km_day',
 'transportation_transfer_station_or_colletion_point_aggregation_na',
 'additional_data_total_swm_expenditures_year',
 'additional_data_total_annual_costs_to_collect_and_dispose_of_city_s_waste_year',
 'additional_data_total_waste_management_budget_year',
 'additional_data_swm_contract_arrangement_1_year_contract_period',
 'additional_data_other_user_fees_na',
 'additional_data_number_of_scavengers_on_dumpsites_number',
 'additional_data_collection_ton',
 'additional_data_annual_waste_budget_year',
 'additional_data_annual_swm_budget_2017_year',
 'waste_treatment_advanced_thermal_treatment_percent',
 'waste_collection_cost_recovery_household_fee_to_private_agents_non_municipal_na',
 'additional_data_annual_sw

In [20]:
cities_df_1 = cities_df.drop(ColToDrop, axis=1)
cities_df_1

iso3c region_id country_name  income_id       city_name  \
0     AFG       SAS  Afghanistan  LowIncome       Jalalabad   
1     AFG       SAS  Afghanistan  LowIncome        Kandahar   
2     AFG       SAS  Afghanistan  LowIncome  Mazar-E-Sharif   
3     AFG       SAS  Afghanistan  LowIncome           Kabul   
4     AFG       SAS  Afghanistan  LowIncome         Hirat    
..    ...       ...          ...        ...             ...   
362   ZWE       SSF     Zimbabwe  LowIncome        Harare    
363   ZWE       SSF     Zimbabwe  LowIncome           Gweru   
364   ZWE       SSF     Zimbabwe  LowIncome          Kariba   
365   ZWE       SSF     Zimbabwe  LowIncome   Masvingo City   
366   ZWE       SSF     Zimbabwe  LowIncome       Sakubva    

     composition_food_organic_waste_percent  composition_glass_percent  \
0                                    1.0455                     1.1175   
1                                   11.6000                     2.3110   
2                                   50.5000                     1.3200   
3                                   13.5000                     1.0000   
4                                   52.9400                     1.4700   
..                                      ...                        ...   
362                                 22.6100                     6.4500   
363                                 62.4000                     3.0000   
364                                 52.0000                     2.0000   
365                                 15.0000                     4.0000   
366                                 32.0000                     5.0000   

     composition_metal_percent  composition_other_percent  \
0                       0.0175                     84.854   
1                       0.1690                     64.364   
2                       0.3200                     28.870   
3                       0.5000                     63.000   
4                       0.2300                     23.050   
..                         ...                        ...   
362                        NaN                     32.740   
363                     3.8000                     13.900   
364                     3.0000                     25.000   
365                     5.0000                      6.000   
366                     6.0000                      7.000   

     composition_paper_cardboard_percent  ...  \
0                                  0.080  ...   
1                                  1.499  ...   
2                                  1.690  ...   
3                                  5.000  ...   
4                                  3.670  ...   
..                                   ...  ...   
362                               19.690  ...   
363                                6.700  ...   
364                                7.000  ...   
365                               30.000  ...   
366                               27.000  ...   

     separation_breakdonw_cans_and_metals_na separation_breakdonw_glass_na  \
0                                         No                            No   
1                                         No                            No   
2                                         No                            No   
3                                        NaN                           NaN   
4                                         No                            No   
..                                       ...                           ...   
362                                      NaN                           NaN   
363                                      NaN                           NaN   
364                                      NaN                           NaN   
365                                      NaN                           NaN   
366                                      NaN                           NaN   

    separation_breakdonw_organics_na separation_breakdonw_paper_cardboard_na  \
0                  

***country_name* variable can be dropped because its information is repeated in *iso3c*.**

***iso3c* is country code according to United Nations (https://unstats.un.org/unsd/tradekb/knowledgebase/country-code).**

In [21]:
cities_df_2 = cities_df_1.drop(['country_name'],axis=1)
cities_df_2 = cities_df_2.rename(columns={'iso3c': 'country_code'})
cities_df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 367 entries, 0 to 366
Data columns (total 28 columns):
country_code                                                                                                     367 non-null object
region_id                                                                                                        367 non-null object
income_id                                                                                                        367 non-null object
city_name                                                                                                        367 non-null object
composition_food_organic_waste_percent                                                                           291 non-null float64
composition_glass_percent                                                                                        269 non-null float64
composition_metal_percent                                                                      

In [22]:
for x in cities_df_2.columns:
    if cities_df_2[x].dtypes==object:
        if x not in ['region_id','country_code','city_name']:
            print (x,' , ', cities_df_2[x].unique())

income_id  ,  ['LowIncome' 'LowMiddleIncome' 'MiddleUpperIncome' 'HighIncome']
institutional_framework_department_dedicated_to_solid_waste_management_na  ,  ['Yes' nan 'No']
institutional_framework_information_system_for_solid_waste_management_na  ,  ['No' nan 'Yes' 'None']
institutional_framework_unit_enforcing_solid_waste_issues_in_the_city_such_as_illegal_dumping_or_littering_na  ,  [nan 'No' 'Yes']
legal_framework_long_term_integrated_solid_waste_master_plan_na  ,  ['No' nan 'yes' 'Yes' 'None']
legal_framework_master_plan_is_being_implemented_na  ,  ['No' nan 'yes' 'Yes']
legal_framework_solid_waste_management_rules_and_regulations_na  ,  ['Yes' nan 'yes' 'No']
primary_collection_mode_form_of_primary_collection_na  ,  ['Door-to-door collection' 'Centralized drop-off point' nan
 'Curbside collection' 'Curbside' 'Multiple'
 'Door-to-door collection: waste collector knocks on door or rings doorbell and waits for resident to bring waste.'
 'Other' 'Door-to-door-collection' 'Block colle

***waste_collection_cost_recovery_household_fee_amount_na* variable should be recognized as numerical value but we do not have information about considered currency and values are very disparate, so we will also drop this variable.**

In [23]:
cities_df_3 = cities_df_2.drop(['waste_collection_cost_recovery_household_fee_amount_na'], axis=1)

#### We notice that among the categorical variables, some have binary variables with NO or YES values but in some cases null information is considered instaed of NO or the variables have unappropriate repeated values because of lower or upper case writing. We correct this:

In [24]:
cities_df_3.replace({'Yes':'yes','No':'no','None':'no'},inplace=True)

In [25]:
VariableReplaceNanbyNo = []
for x in cities_df_3.columns:
    if cities_df_3[x].dtypes==object:
        if 'yes' in cities_df_3[x].unique():
            VariableReplaceNanbyNo.append(x)

VariableReplaceNanbyNo            

['institutional_framework_department_dedicated_to_solid_waste_management_na',
 'institutional_framework_information_system_for_solid_waste_management_na',
 'institutional_framework_unit_enforcing_solid_waste_issues_in_the_city_such_as_illegal_dumping_or_littering_na',
 'legal_framework_long_term_integrated_solid_waste_master_plan_na',
 'legal_framework_master_plan_is_being_implemented_na',
 'legal_framework_solid_waste_management_rules_and_regulations_na',
 'separation_breakdonw_cans_and_metals_na',
 'separation_breakdonw_glass_na',
 'separation_breakdonw_organics_na',
 'separation_breakdonw_paper_cardboard_na',
 'separation_breakdonw_plastics_and_packaging_na',
 'separation_existence_of_source_separation_na']

In [26]:
for x in VariableReplaceNanbyNo:
    cities_df_3[x].fillna('no', inplace=True)

cities_df_3.isna().sum()

country_code                                                                                                       0
region_id                                                                                                          0
income_id                                                                                                          0
city_name                                                                                                          0
composition_food_organic_waste_percent                                                                            76
composition_glass_percent                                                                                         98
composition_metal_percent                                                                                        102
composition_other_percent                                                                                         77
composition_paper_cardboard_percent                             

In [27]:
for x in cities_df_3.columns:
    if cities_df_3[x].dtypes==object:
        if x not in ['region_id','country_code','city_name']:
            print (x,' , ', cities_df_3[x].unique())

income_id  ,  ['LowIncome' 'LowMiddleIncome' 'MiddleUpperIncome' 'HighIncome']
institutional_framework_department_dedicated_to_solid_waste_management_na  ,  ['yes' 'no']
institutional_framework_information_system_for_solid_waste_management_na  ,  ['no' 'yes']
institutional_framework_unit_enforcing_solid_waste_issues_in_the_city_such_as_illegal_dumping_or_littering_na  ,  ['no' 'yes']
legal_framework_long_term_integrated_solid_waste_master_plan_na  ,  ['no' 'yes']
legal_framework_master_plan_is_being_implemented_na  ,  ['no' 'yes']
legal_framework_solid_waste_management_rules_and_regulations_na  ,  ['yes' 'no']
primary_collection_mode_form_of_primary_collection_na  ,  ['Door-to-door collection' 'Centralized drop-off point' nan
 'Curbside collection' 'Curbside' 'Multiple'
 'Door-to-door collection: waste collector knocks on door or rings doorbell and waits for resident to bring waste.'
 'Other' 'Door-to-door-collection' 'Block collection'
 'centralized and curbside' 'Door to door collect

In [28]:
cities_df_3['primary_collection_mode_form_of_primary_collection_na'].fillna('not_informed', inplace=True)

In [29]:
cities_df_3['primary_collection_mode_form_of_primary_collection_na'].unique()

array(['Door-to-door collection', 'Centralized drop-off point',
       'not_informed', 'Curbside collection', 'Curbside', 'Multiple',
       'Door-to-door collection: waste collector knocks on door or rings doorbell and waits for resident to bring waste.',
       'Other', 'Door-to-door-collection', 'Block collection',
       'centralized and curbside', 'Door to door collection'],
      dtype=object)

In [30]:
cities_df_3['primary_collection_mode_form_of_primary_collection_na']=cities_df_3['primary_collection_mode_form_of_primary_collection_na'].str.lower()
cities_df_3['primary_collection_mode_form_of_primary_collection_na'].unique()

array(['door-to-door collection', 'centralized drop-off point',
       'not_informed', 'curbside collection', 'curbside', 'multiple',
       'door-to-door collection: waste collector knocks on door or rings doorbell and waits for resident to bring waste.',
       'other', 'door-to-door-collection', 'block collection',
       'centralized and curbside', 'door to door collection'],
      dtype=object)

In [31]:
cities_df_3['primary_collection_mode_form_of_primary_collection_na'].loc[cities_df_3['primary_collection_mode_form_of_primary_collection_na'].str.contains('door')] = 'door to door'
cities_df_3['primary_collection_mode_form_of_primary_collection_na'].unique()

/Users/marike/opt/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:205: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



array(['door to door', 'centralized drop-off point', 'not_informed',
       'curbside collection', 'curbside', 'multiple', 'other',
       'block collection', 'centralized and curbside'], dtype=object)

In [32]:
cities_df_3['primary_collection_mode_form_of_primary_collection_na'].loc[cities_df_3['primary_collection_mode_form_of_primary_collection_na'].str.contains('curbside')] = 'curbside'
cities_df_3['primary_collection_mode_form_of_primary_collection_na'].unique()

array(['door to door', 'centralized drop-off point', 'not_informed',
       'curbside', 'multiple', 'other', 'block collection'], dtype=object)

In [33]:
cities_df_3.isna().sum()

country_code                                                                                                       0
region_id                                                                                                          0
income_id                                                                                                          0
city_name                                                                                                          0
composition_food_organic_waste_percent                                                                            76
composition_glass_percent                                                                                         98
composition_metal_percent                                                                                        102
composition_other_percent                                                                                         77
composition_paper_cardboard_percent                             

Some columns name are very long and many correspond to subcategories. So for better reading of the information in this notebook, we will modify them using the following expressions:

***compo*** for composition (waste type) percent
    
***IF*** for Institutional Frame work
    
***SW*** for Solid Waste
    
***SW*** for Solid Waste, ***mgmt*** for management
    
***dept*** for department
    
***info*** for information
    
***LF*** for Legal Frame
    
***Separation*** for Separation Breakdown

In [34]:
cities_df_3 = cities_df_3.rename(columns={
'composition_food_organic_waste_percent':'compo_organic', 
'composition_glass_percent':'compo_glass',
'composition_metal_percent':'compo_metal',
'composition_other_percent':'compo_other',
'composition_paper_cardboard_percent':'compo_paper_cardboard',
'composition_plastic_percent':'compo_plastic',
'institutional_framework_department_dedicated_to_solid_waste_management_na':'IF_SW_mgmt_dept',
'institutional_framework_information_system_for_solid_waste_management_na':'IF_SW_mgmt_info',
'institutional_framework_unit_enforcing_solid_waste_issues_in_the_city_such_as_illegal_dumping_or_littering_na':'IF_SW_control_enforcing_unit',
'legal_framework_long_term_integrated_solid_waste_master_plan_na':'LF_long_term_integrated_SW_master_plan',
'legal_framework_master_plan_is_being_implemented_na':'LF_SW_master_plan_being_implemented',
'legal_framework_solid_waste_management_rules_and_regulations_na':'LF_SW_mgmt_rules_and_regulations',
'population_population_number_of_people':'population',
'primary_collection_mode_form_of_primary_collection_na':'primary_collection_form',
'separation_breakdonw_cans_and_metals_na':'separation_metals',
'separation_breakdonw_glass_na':'separation_glass',
'separation_breakdonw_organics_na':'separation_organic',
'separation_breakdonw_paper_cardboard_na':'separation_paper_cardboard',
'separation_breakdonw_plastics_and_packaging_na':'separation_plastics_packaging',
'separation_existence_of_source_separation_na':'separation_existence',
'total_msw_total_msw_generated_tons_year':'MSW_generated_tons_year',
'waste_collection_coverage_total_percent_of_waste_percent_of_waste':'waste_collection_coverage_percent', 
'waste_treatment_recycling_percent':'recycling_percent'
})                                        

We are very interested in recycling information when it is done but since the information is missing for 239 cities of the dataset, we extract the corresponding rows in a specific dataset that we keep to analyse more ahead in the project and we drop this variable from our main dataset.

In [35]:
cities_df_4 = cities_df_3
cities_df_4[['waste_collection_coverage_percent','recycling_percent']]=cities_df_4[['waste_collection_coverage_percent','recycling_percent']].fillna(value=0)
cities_df_4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 367 entries, 0 to 366
Data columns (total 27 columns):
country_code                              367 non-null object
region_id                                 367 non-null object
income_id                                 367 non-null object
city_name                                 367 non-null object
compo_organic                             291 non-null float64
compo_glass                               269 non-null float64
compo_metal                               265 non-null float64
compo_other                               290 non-null float64
compo_paper_cardboard                     290 non-null float64
compo_plastic                             283 non-null float64
IF_SW_mgmt_dept                           367 non-null object
IF_SW_mgmt_info                           367 non-null object
IF_SW_control_enforcing_unit              367 non-null object
LF_long_term_integrated_SW_master_plan    367 non-null object
LF_SW_master_plan_bei

In [36]:
cities_df_4.loc[cities_df_4.population.isna()]

country_code region_id          income_id      city_name  compo_organic  \
15           ASM       EAS  MiddleUpperIncome      Pago Pago           19.7   
60           CMR       SSF    LowMiddleIncome          Limbe           54.8   
117          GNQ       SSF  MiddleUpperIncome           Labe           69.0   
250          NGA       SSF    LowMiddleIncome           Kano           43.0   
292          PSE       MEA    LowMiddleIncome       Ramallah            NaN   
353          ZAF       SSF  MiddleUpperIncome         Soweto            9.0   
358          ZWE       SSF          LowIncome    Kadoma City           60.0   
359          ZWE       SSF          LowIncome          Mbare           28.0   
365          ZWE       SSF          LowIncome  Masvingo City           15.0   
366          ZWE       SSF          LowIncome      Sakubva            32.0   

     compo_glass  compo_metal  compo_other  compo_paper_cardboard  \
15           3.4          7.9         25.6                   26.4   
60           1.6          2.4         16.3                   12.5   
117          0.3          1.4          2.0                    4.1   
250          2.0          5.0         29.0                   17.0   
292          NaN          NaN          NaN                    NaN   
353         12.0          3.0         64.0                    9.0   
358          3.0          3.0         22.0                    5.0   
359          2.0         12.0         25.0                    6.0   
365          4.0          5.0          6.0                   30.0   
366          5.0          6.0          7.0                   27.0   

     compo_plastic  ... primary_collection_form separation_metals  \
15            12.8  ...            not_informed                no   
60            12.4  ...            not_informed                no   
117           22.8  ...            not_informed                no   
250            4.0  ...            not_informed                no   
292            NaN  ...            not_informed                no   
353            3.0  ...            not_informed                no   
358            7.0  ...            not_informed                no   
359           11.0  ...            not_informed                no   
365           40.0  ...            not_informed                no   
366           23.0  ...            not_informed                no   

    separation_glass separation_organic separation_paper_cardboard  \
15                no                 no                         no   
60                no                 no                         no   
117               no                 no                         no   
250               no                 no                         no   
292               no                 no                         no   
353               no                 no                         no   
358               no                 no                         no   
359               no                 no                         no   
365               no                 no                         no   
366               no                 no                         no   

    separation_plastics_packaging  separation_existence  \
15                             no                    no   
60                             no                    no   
117                            no                    no   
250                            no                    no   
292                            no                    no   
353                            no                    no   
358                            no                    no   
359                            no                    no   
365                            no                    no   
366                            no                    no   

    MSW_generated_tons_year waste_collection_coverage_percent  \
15                      NaN                               0.0   
60                      NaN                               0.0  

#### Rows were population number is not informed also correspond to cities where the quantity of generated waste is not indicated. Since the generation of waste is our target variable, we will drop the rows where this information is missing. 

In [37]:
cities_df_4.loc[cities_df_4['MSW_generated_tons_year'].isna()]

country_code region_id          income_id       city_name  compo_organic  \
15           ASM       EAS  MiddleUpperIncome       Pago Pago          19.70   
16           AUS       EAS         HighIncome           Perth          40.10   
17           AUS       EAS         HighIncome       Melbourne          68.21   
36           BGR       ECS  MiddleUpperIncome            Ruse            NaN   
37           BGR       ECS  MiddleUpperIncome          Burgas            NaN   
38           BHR       MEA         HighIncome          Manama            NaN   
56           CHE       ECS         HighIncome          Zurich            NaN   
60           CMR       SSF    LowMiddleIncome           Limbe          54.80   
69           COM       SSF          LowIncome          Moroni            NaN   
73           CYP       ECS         HighIncome      Paralimini          20.00   
77           DEU       ECS         HighIncome         Hamburg          21.49   
89           EST       ECS         HighIncome       Tartummaa           0.93   
100          ETH       SSF          LowIncome   Halaba Kulito          88.80   
102          FIN       ECS         HighIncome            Oulu            NaN   
103          FIN       ECS         HighIncome       Riihimaki          48.40   
104          FIN       ECS         HighIncome           Turku          42.80   
112          GEO       ECS    LowMiddleIncome         Tbilisi          69.50   
117          GNQ       SSF  MiddleUpperIncome            Labe          69.00   
128          HUN       ECS         HighIncome        Debrecen          36.95   
131          IMN       ECS         HighIncome         Douglas            NaN   
182          ISR       MEA         HighIncome       Jerusalem            NaN   
183          ISR       MEA         HighIncome        Tel Aviv            NaN   
194          KAZ       ECS  MiddleUpperIncome        Kostanay            NaN   
195          KEN       SSF    LowMiddleIncome       Eldoret             NaN   
202          KWT       MEA         HighIncome     Kuwait City            NaN   
216          LVA       ECS         HighIncome         Liepaja          26.00   
242          MRT       SSF    LowMiddleIncome      Nouakchott           5.00   
248          NGA       SSF    LowMiddleIncome          Ilorin          61.00   
250          NGA       SSF    LowMiddleIncome            Kano          43.00   
256          NLD       ECS         HighIncome       Eindhoven          39.00   
292          PSE       MEA    LowMiddleIncome        Ramallah            NaN   
300          RWA       SSF          LowIncome       Nyagatare            NaN   
330          TUV       EAS  MiddleUpperIncome        Funafuti          52.40   
341          UZB       ECS    LowMiddleIncome        Namangan            NaN   
347          WSM       EAS  MiddleUpperIncome    Upolu (Apia)           3.80   
353          ZAF       SSF  MiddleUpperIncome          Soweto           9.00   
354          ZAF       SSF  MiddleUpperIncome  Bloemfontein           42.60   
358          ZWE       SSF          LowIncome     Kadoma City          60.00   
359          ZWE       SSF          LowIncome           Mbare          28.00   
365          ZWE       SSF          LowIncome   Masvingo City          15.00   
366          ZWE       SSF          LowIncome       Sakubva           32.00   

     compo_glass  compo_metal  compo_other  compo_paper_cardboard  \
15          3.40         7.90        25.60                  26.40   
16          6.80         3.50        22.00                  15.20   
17          3.00          NaN         1.00                  11.44   
36           NaN          NaN          NaN                    NaN   
37           NaN          NaN          NaN                    NaN   
38           NaN          NaN          NaN                    NaN   
56           NaN          NaN          NaN                    NaN   
60          1.60         2.40        16.30                  12.50   
69           NaN      

In [38]:
cities_df_5 = cities_df_4.dropna(subset = ['MSW_generated_tons_year']) 

In [39]:
cities_df_5.isna().sum()

country_code                               0
region_id                                  0
income_id                                  0
city_name                                  0
compo_organic                             61
compo_glass                               81
compo_metal                               84
compo_other                               62
compo_paper_cardboard                     62
compo_plastic                             68
IF_SW_mgmt_dept                            0
IF_SW_mgmt_info                            0
IF_SW_control_enforcing_unit               0
LF_long_term_integrated_SW_master_plan     0
LF_SW_master_plan_being_implemented        0
LF_SW_mgmt_rules_and_regulations           0
population                                 0
primary_collection_form                    0
separation_metals                          0
separation_glass                           0
separation_organic                         0
separation_paper_cardboard                 0
separation

We create an independent dataset for further study with the information of waste composition when these data are given.

In [40]:
waste_composition = []
for x in cities_df_5.columns:
    if 'compo' in x:
        waste_composition.append(x)
waste_composition        

['compo_organic',
 'compo_glass',
 'compo_metal',
 'compo_other',
 'compo_paper_cardboard',
 'compo_plastic']

In [41]:
waste_composition_df = cities_df_5.dropna(subset = waste_composition) 
print(waste_composition_df.shape)
waste_composition_df.info()

(225, 27)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 225 entries, 0 to 364
Data columns (total 27 columns):
country_code                              225 non-null object
region_id                                 225 non-null object
income_id                                 225 non-null object
city_name                                 225 non-null object
compo_organic                             225 non-null float64
compo_glass                               225 non-null float64
compo_metal                               225 non-null float64
compo_other                               225 non-null float64
compo_paper_cardboard                     225 non-null float64
compo_plastic                             225 non-null float64
IF_SW_mgmt_dept                           225 non-null object
IF_SW_mgmt_info                           225 non-null object
IF_SW_control_enforcing_unit              225 non-null object
LF_long_term_integrated_SW_master_plan    225 non-null object
LF_SW_maste

**COMPOSITION OF WASTE IN FUNCTION OF GENERATED WASTE & INCOME**

Since we have the waste composition dataset without null information about waste composition, we drop the rows in the recycling dataset where waste composition information is missing.

In [42]:
"""Finally we create one CSV files for further study if requested: 
with complete information regarding cities waste composition (25 variables, 225 cities)."""

waste_composition_df.to_csv('CitiesWasteComposition.csv', index=False)

In [43]:
#finally we create our cities dataset where target variable we are interested in is waste generation, without null information.
cities_df_final = cities_df_5.drop(waste_composition, axis = 1) 

In [44]:
cities_df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 326 entries, 0 to 364
Data columns (total 21 columns):
country_code                              326 non-null object
region_id                                 326 non-null object
income_id                                 326 non-null object
city_name                                 326 non-null object
IF_SW_mgmt_dept                           326 non-null object
IF_SW_mgmt_info                           326 non-null object
IF_SW_control_enforcing_unit              326 non-null object
LF_long_term_integrated_SW_master_plan    326 non-null object
LF_SW_master_plan_being_implemented       326 non-null object
LF_SW_mgmt_rules_and_regulations          326 non-null object
population                                326 non-null float64
primary_collection_form                   326 non-null object
separation_metals                         326 non-null object
separation_glass                          326 non-null object
separation_organic        

In [45]:
print('Our final cleaned Cities dataset for the prediction of waste generation has the following shape: ',cities_df_final.shape,'.')

Our final cleaned Cities dataset for the prediction of waste generation has the following shape:  (326, 21) .


#### Finally after exporting our subdatset one with waste composition , we obtain our cleaned final dataset based on 326 cities and 21 variables that we export as CSV file.

In [46]:
cities_df_final.to_csv('CitiesWasteGeneration.csv', index=False)